sql: prevent gateway from always being picked as the default

adityamaru opened the pull request:
Previously, the instance resolver would always
assign the partition span to the gateway if the
gateway was in the set of eligible instances and
we did not find an eligible instance with a better
locality match. In large clusters during backup/cdc
running with execution locality, this could cause
the gateway to get the lions share of work thereby
causing it to OOM or severely throttle performance.

This change make span partitioning a little more
stateful. Concretely, we now track how many partition
spans have been assigned to each node in the `planCtx`
that is used throughout the planning of a single statement.
This distribution is then used to limit the number of
partition spans we default to the gateway. Currently, by
default we allow the gateway to have:

`2 * average number of partition spans across the other instances`

If the gateway does not satisfy this heuristic we randomly
pick one of the other eligible instances. Note, if there
are no eligible instances except for the gateway, or the
gateway has received no spans yet, we will pick the gateway.

Fixes: https://github.com/cockroachdb/cockroach/issues/114079
Release note (bug fix): fixes a bug where large jobs running
with execution locality could result in the gateway being assigned
most of the work causing performance degradation and cluster
instability

adityamaru committed:
sql: prevent gateway from always being picked

Previously, the instance resoler would always
assign the partition span to the gateway if the
gateway was in the set of eligible instances and
we did not find an eligible instance with a better
locality match. In large clusters during backup/cdc
running with execution locality, this could cause
the gateway to get the lions share of work thereby
causing it to OOM or severely throttle performance.

This change make span partitioning a little more
stateful. Concretely, we now track how many partition
spans have been assigned to each node in the `planCtx`
that is used throughout the planning of a single statement.
This distribution is then used to limit the number of
partition spans we default to the gateway. Currently, by
default we allow the gateway to have:

`2 * average number of partition spans across the other instances`

If the gateway does not satisfy this heuristic we randomly
pick one of the other eligible instances. Note, if there
are no eligible instances except for the gateway, or the
gateway has received no spans yet, we will pick the gateway.

Fixes: #114079
Release note (bug fix): fixes a bug where large jobs running
with execution locality could result in the gateway being assigned
most of the work causing performance degradation and cluster
instability

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114537)
<!-- Reviewable:end -->

adityamaru committed:
sql: add `distsql_plan_gateway_bias` session var

This change adds a new session variable to control
how many times the gateway will be picked as the default
target for a partition relative to the distribution of
partition spans across other nodes.

Release note: None

yuzefovich reviewed (commented):
Looks good, I have some comments.

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-:-NjPaepFCxnZu3krW6bi:bxyxgod)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @rharding6373)

___
*[`pkg/sql/distsql_physical_planner.go` line 806 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPXFKECoVi77ERuIzr:-NjPXFKECoVi77ERuIzs:b-e0sg9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L806)):*
> ```Go
> // SpanPartitionState captures information about the current state of the
> // partitioning that has occurred during the planning process.
> type SpanPartitionState struct {
> ```

nit: does this need to be exported?
___
*[`pkg/sql/distsql_physical_planner.go` line 809 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPX_LK--sL3pj_K3pg:-NjPX_LK--sL3pj_K3ph:b-t888in) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L809)):*
> ```Go
> 	// partitionSpanDecisions is a mapping from a SpanPartitionReason to the number of
> 	// times we have picked an instance for that reason.
> 	partitionSpanDecisions map[SpanPartitionReason]int
> ```

Given that we have a limited number of `SpanPartitionReason`s, we could allocate fixed-size array instead of the map. We'd probably need to add a sentinel value to `SpanPartitionReason` enum to indicate the current count.
___
*[`pkg/sql/distsql_physical_planner.go` line 823 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPYTNLDx81L8hYr_E2:-NjPYTNLDx81L8hYr_E3:bwl9feo) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L823)):*
> ```Go
> 
> // Update updates the SpanPartitionState with the information about the new span partition.
> func (p *SpanPartitionState) Update(partitionNode base.SQLInstanceID, partitionReason SpanPartitionReason) {
> ```

nit: ditto for unexporting.
___
*[`pkg/sql/distsql_physical_planner.go` line 838 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPYbz68pgEjh-2xefW:-NjPYbz7-rlfex55w6Wt:b-m33vc) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L838)):*
> ```Go
> 	// spanPartitionState captures information about the current state of the
> 	// partitioning that has occurred during the planning process.
> 	spanPartitionState *SpanPartitionState
> ```

nit: do we need this to be a pointer?
___
*[`pkg/sql/distsql_physical_planner.go` line 1143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPYfCt3bsBDhekYhwI:-NjPYfCt3bsBDhekYhwJ:b-rgogh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L1143)):*
> ```Go
> 	// on any of the instances and so we default to a naive round-robin strategy.
> 	SpanPartitionReason_ROUND_ROBIN SpanPartitionReason = 8
> ```

nit: redundant empty line.
___
*[`pkg/sql/distsql_physical_planner.go` line 1576 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPYvUM1pgUR74xekvX:-NjPYvUM1pgUR74xekvY:b-u0xdku) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L1576)):*
> ```Go
> // shouldPickGateway determines whether the gateway node should be picked for a
> // particular partition.
> func (dsp *DistSQLPlanner) shouldPickGateway(_ context.Context, planCtx *PlanningCtx, instances []sqlinstance.InstanceInfo) bool {
> ```

nit: why pass context if we're not using it?
___
*[`pkg/sql/distsql_physical_planner.go` line 1591 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPa5Lc8QtEnDMTlt8N:-NjPa5Lc8QtEnDMTlt8O:b-hh61d0) ([raw file](https://github.com/cockroachdb/cockroach/blob/217cd5cb2aae070f3bc85e172beb3a1ffa7229a5/pkg/sql/distsql_physical_planner.go#L1591)):*
> ```Go
> 	}
> 
> 	// If the gateway has span partitions >= twice the average span partitions
> ```

nit: maybe do `s/twice/twice (by default)/`.
___
*[`pkg/sql/vars.go` line 3103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPaDs0BZ0Vn-O08a85:-NjPaDs0BZ0Vn-O08a86:b-ymlsjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/217cd5cb2aae070f3bc85e172beb3a1ffa7229a5/pkg/sql/vars.go#L3103)):*
> ```Go
> 				return err
> 			}
> 			m.SetDistSQLPlanGatewayBias(i)
> ```

We should add some validation here, like `i` must be positive or non-negative.
___
*[`pkg/sql/sessiondatapb/session_data.proto` line 139 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPaM8x-QC7rvt-tenW:-NjPaM8x-QC7rvt-tenX:bb4fcvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/217cd5cb2aae070f3bc85e172beb3a1ffa7229a5/pkg/sql/sessiondatapb/session_data.proto#L139)):*
> ```proto
>   // choice relative to the distribution of partition spans on other eligible
>   // instances. Refer to `shouldPickGateway` for more details.
>   int64 distsql_plan_gateway_bias = 31;
> ```

Does this need to be propagated to the remote nodes? I think we make physical planning decisions fully on the gateway, so this should go into `LocalOnlySessionData`.
___
*[`pkg/sql/distsql_physical_planner_test.go` line 1318 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjP_njgA4Z7oxNGFIfK:-NjP_njgA4Z7oxNGFIfL:bibo2fm) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner_test.go#L1318)):*
> ```Go
> 				3: 0,
> 			}},
> 			shouldPick: false,
> ```

Doesn't the name of this test case "below threshold" imply that we should pick the gateway?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114537)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @rharding6373)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPpmuRBQ6ZA0fr6F3l):*
BTW why did we need this plumbing? On a quick glance we should have access to `planCtx` in all relevant places already (although via an "outer variable" capture).


<!-- Sent from Reviewable.io -->

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114537)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @rharding6373, and @yuzefovich)

___
*[`pkg/sql/distsql_physical_planner.go` line 806 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPXFKECoVi77ERuIzr:-NjmZxv30Zym-j2fx05l:b-xks305) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L806)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: does this need to be exported?
</blockquote></details>

In a follow up I want jobs to persist some of this SpanPartitionState in the job_info table and annotate their distsql diagrams with it. I've unexported it for now but I suspect I will change it back soon.
___
*[`pkg/sql/distsql_physical_planner.go` line 809 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPX_LK--sL3pj_K3pg:-Njmkl30-sifbQqN42vm:b-21w2pp) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L809)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Given that we have a limited number of `SpanPartitionReason`s, we could allocate fixed-size array instead of the map. We'd probably need to add a sentinel value to `SpanPartitionReason` enum to indicate the current count.
</blockquote></details>

good idea, done.
___
*[`pkg/sql/distsql_physical_planner.go` line 823 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPYTNLDx81L8hYr_E2:-Njm_HdSB6xCbCpHnWTR:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L823)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: ditto for unexporting.
</blockquote></details>

done
___
*[`pkg/sql/distsql_physical_planner.go` line 838 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPYbz68pgEjh-2xefW:-Njmly26-auBXPRo3ewB:b-9gjiqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L838)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: do we need this to be a pointer?
</blockquote></details>

I made it a pointer because we have a `totalPartitions` int in the struct and if we pass the *planCtx into a method that mutates the partitionState.counter it should reflect at the callsite too. Let me know if I'm missing something.
___
*[`pkg/sql/vars.go` line 3103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPaDs0BZ0Vn-O08a85:-Njn1g2O-UGoN2XbPVTf:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/217cd5cb2aae070f3bc85e172beb3a1ffa7229a5/pkg/sql/vars.go#L3103)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We should add some validation here, like `i` must be positive or non-negative.
</blockquote></details>

done
___
*[`pkg/sql/sessiondatapb/session_data.proto` line 139 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPaM8x-QC7rvt-tenW:-NjmeFyP5SqqP5pNODEV:b-xvc14z) ([raw file](https://github.com/cockroachdb/cockroach/blob/217cd5cb2aae070f3bc85e172beb3a1ffa7229a5/pkg/sql/sessiondatapb/session_data.proto#L139)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Does this need to be propagated to the remote nodes? I think we make physical planning decisions fully on the gateway, so this should go into `LocalOnlySessionData`.
</blockquote></details>

I thought about this, the reason i added it to SessionData is that jobs know how to shuttle our SessionData around https://github.com/cockroachdb/cockroach/blob/aeec29996e7199575d7585a7851756d2331a81d2/pkg/jobs/jobspb/jobs.proto#L912 in its details. Currently, we don't teach backup/restore/import etc. to respect the session vars set in this serialized session data but in the future I could see us doing so. A job can be coordinated by any node in the cluster so its not guaranteed to be the gateway.
___
*[`pkg/sql/distsql_physical_planner_test.go` line 1318 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjP_njgA4Z7oxNGFIfK:-Njn4OK580hUY4PniUCR:b-x06mkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner_test.go#L1318)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Doesn't the name of this test case "below threshold" imply that we should pick the gateway?
</blockquote></details>

you're right, in this case we will not pick the gateway until the other nodes have atleast 4 partitions assigned to them. I have renamed the test but let me know if this behaviour is strange to you.


<!-- Sent from Reviewable.io -->

adityamaru commented:
> BTW why did we need this plumbing? On a quick glance we should have access to planCtx in all relevant places already (although via an "outer variable" capture).

I've addressed everything except this. Looking into if I can drop that commit but this should be rfal.

adityamaru commented:
> BTW why did we need this plumbing? On a quick glance we should have access to planCtx in all relevant places already (although via an "outer variable" capture).

nice callout, i could drop the first commit.

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r7, 7 of 7 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-:-NjoSRJYAoYVm9VrKCuj:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @rharding6373)

___
*[`pkg/sql/distsql_physical_planner.go` line 806 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPXFKECoVi77ERuIzr:-NjoN_FZ95GLtzidHu6T:bce10or) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L806)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

In a follow up I want jobs to persist some of this SpanPartitionState in the job_info table and annotate their distsql diagrams with it. I've unexported it for now but I suspect I will change it back soon.
</blockquote></details>

Oh, I see, then it's up to you if know that it'll be used outside of `pkg/sql`.
___
*[`pkg/sql/distsql_physical_planner.go` line 838 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPYbz68pgEjh-2xefW:-NjoPG8g2nARofI8B_RS:b-d3yjyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/145624327df9e14ccf451836d9214d60ac8bf672/pkg/sql/distsql_physical_planner.go#L838)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I made it a pointer because we have a `totalPartitions` int in the struct and if we pass the *planCtx into a method that mutates the partitionState.counter it should reflect at the callsite too. Let me know if I'm missing something.
</blockquote></details>

I think because we always access it via `planCtx.spanPartitionState.*`, we _could_ have it stored as a value. Like `update` method has pointer receiver, etc. Please feel free to ignore this comment - it's a "nit" for a reason ;)
___
*[`pkg/sql/distsql_physical_planner.go` line 1153 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjoQafL6czpMtbmrnx9:-NjoQafM-z5ZI9l36aOj:b4zuhmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/472ee57813f3e33df6df4a2910ab7a1a1d359f82/pkg/sql/distsql_physical_planner.go#L1153)):*
> ```Go
> 	// SpanPartitionReason_MAX_VALUE is the maximum number of
> 	// SpanPartitionReasons.
> 	SpanPartitionReason_MAX_VALUE SpanPartitionReason = 12
> ```

nit: you're probably aware of this, but we could have `SpanPartitionReason_UNSPECIFIED SpanPartitionReason = iota`, and then we wouldn't have to explicitly specify the value for each. This would require that `SpanPartitionReason_LOCALITY_AWARE_RANDOM_GATEWAY_OVERLOADED` is added last, but the upside is that we wouldn't have to update `SpanPartitionReason_MAX_VALUE` in case we add another reason in the future.
___
*[`pkg/sql/distsql_physical_planner.go` line 1700 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjoR6De-kYcGuohIGyl:-NjoR6De-kYcGuohIGym:b-k32e4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/472ee57813f3e33df6df4a2910ab7a1a1d359f82/pkg/sql/distsql_physical_planner.go#L1700)):*
> ```Go
> 
> 			// No instances had any locality tiers in common with the node locality so
> 			// just return the gateway if it is eligible. If it isn't, just pick a
> ```

nit: this comment needs an update.
___
*[`pkg/sql/vars.go` line 3129 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjoOLTp2mY_C_5KaXrR:-NjoOLTp2mY_C_5KaXrS:b-d05l2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8399f4a9448e3f7ec723cdba742ca4e0b627e8a/pkg/sql/vars.go#L3129)):*
> ```Go
> 			if i < 1 {
> 				return pgerror.Newf(pgcode.InvalidParameterValue,
> 					"cannot set distsql_plan_gateway_bias to a negative value: %d", i)
> ```

nit: `s/negative/non-positive/` to be precise :)
___
*[`pkg/sql/sessiondatapb/session_data.proto` line 139 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPaM8x-QC7rvt-tenW:-NjoOS31048rxTwTr15A:b3scdhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/217cd5cb2aae070f3bc85e172beb3a1ffa7229a5/pkg/sql/sessiondatapb/session_data.proto#L139)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I thought about this, the reason i added it to SessionData is that jobs know how to shuttle our SessionData around https://github.com/cockroachdb/cockroach/blob/aeec29996e7199575d7585a7851756d2331a81d2/pkg/jobs/jobspb/jobs.proto#L912 in its details. Currently, we don't teach backup/restore/import etc. to respect the session vars set in this serialized session data but in the future I could see us doing so. A job can be coordinated by any node in the cluster so its not guaranteed to be the gateway.
</blockquote></details>

Oh, interesting, I wasn't aware of this. Consider mentioning this detail in the comment (since I think pretty much all other fields in this struct affect `EvalCtx` which affects execution, and this one seems like an exception).


<!-- Sent from Reviewable.io -->

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114537)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @rharding6373, and @yuzefovich)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPpmuRBQ6ZA0fr6F3l):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

BTW why did we need this plumbing? On a quick glance we should have access to `planCtx` in all relevant places already (although via an "outer variable" capture).
</blockquote></details>

resolved by dropping the commit
___
*[`pkg/sql/distsql_physical_planner.go` line 1153 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjoQafL6czpMtbmrnx9:-Njrt4xu1fhMpsQg59Ul:b2zrnka) ([raw file](https://github.com/cockroachdb/cockroach/blob/472ee57813f3e33df6df4a2910ab7a1a1d359f82/pkg/sql/distsql_physical_planner.go#L1153)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: you're probably aware of this, but we could have `SpanPartitionReason_UNSPECIFIED SpanPartitionReason = iota`, and then we wouldn't have to explicitly specify the value for each. This would require that `SpanPartitionReason_LOCALITY_AWARE_RANDOM_GATEWAY_OVERLOADED` is added last, but the upside is that we wouldn't have to update `SpanPartitionReason_MAX_VALUE` in case we add another reason in the future.
</blockquote></details>

makes sense, changed to iota
___
*[`pkg/sql/sessiondatapb/session_data.proto` line 139 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114537#-NjPaM8x-QC7rvt-tenW:-NjrngHG6DjBsxqLPOpy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/217cd5cb2aae070f3bc85e172beb3a1ffa7229a5/pkg/sql/sessiondatapb/session_data.proto#L139)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Oh, interesting, I wasn't aware of this. Consider mentioning this detail in the comment (since I think pretty much all other fields in this struct affect `EvalCtx` which affects execution, and this one seems like an exception).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

adityamaru commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -1635,10 +1699,24 @@ func (dsp *DistSQLPlanner) makeInstanceResolver(
+					return planCtx.spanPartitionState.testingOverrideRandomSelection(),
+						SpanPartitionReason_LOCALITY_AWARE_RANDOM_GATEWAY_OVERLOADED
+				}
+				for {
```
i think i agree with you here. I did grimace writing the loop for a second so we should follow that intuition and err on the side of picking the gateway.

dt replied:
to be clear: on the side of picking a random node, whether or not it is the gateway. The property we want is that it is uniform and even, not that it _isn't the gateway_, so random (including gateway) seems fine.

adityamaru commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -1103,17 +1132,25 @@ const (
+	// when there is no match to the provided locality filter and the gateway is
+	// eligible but overloaded with other partitions. In this case we pick a
+	// random instance apart from the gateway.
+	SpanPartitionReason_LOCALITY_AWARE_RANDOM_GATEWAY_OVERLOADED SpanPartitionReason = 11
```
I renamed both the enum values to FILTERED instead of AWARE.

adityamaru commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -1534,6 +1574,30 @@ var noInstancesMatchingLocalityFilterErr = errors.New(
+		(planCtx.spanPartitionState.totalPartitionSpans - partitionsOnGateway) / numEligibleInstancesExcludingGateway
+
+	// If the gateway has no span partitions, we should use the gateway.
+	if partitionsOnGateway == 0 {
```
done!

adityamaru commented:
TFTR!

bors r+
